verify/cert.go

```diff
@@ -49,40 +51,60 @@ func NewCertVerifier(cfg *Config, l log.Logger) (*CertVerifier, error) {
 ) error {
-	blockNumber, err := cv.getConfDeepBlockNumber()
+	// 1. verify batch is actually onchain at the batchMetadata's state confirmedBlockNumber
+	// This assert is technically not necessary, but it's a good sanity check.
+	// It could technically happen that a confirmed batch's block gets reorged out,
+	// yet the tx is included in an earlier or later block, making this check fail...
+	confirmationBlockNumber := batchMetadata.GetConfirmationBlockNumber()
+	confirmationBlockNumberBigInt := big.NewInt(0).SetInt64(int64(confirmationBlockNumber))
+	_, err := cv.retrieveBatchMetadataHash(ctx, batchID, confirmationBlockNumberBigInt)
 	if err != nil {
-		return fmt.Errorf("failed to get context block: %w", err)
+		return fmt.Errorf("batch not found onchain at supposedly confirmed block %d: %w", confirmationBlockNumber, err)
 	}
 
-	// 1. ensure that a batch hash can be looked up for a batch ID for a given block number
-	expectedHash, err := cv.manager.BatchIdToBatchMetadataHash(&bind.CallOpts{BlockNumber: blockNumber}, id)
+	// 2. verify that the confirmation status has been reached
+	// We retry 5 times waiting 12 seconds (block time) between each retry,
+	// in case our eth node is behind that of the eigenda_client's node that deemed the batch confirmed
+	onchainHash, err := retry.Do(ctx, 5, retry.Fixed(12*time.Second), func() ([32]byte, error) {
+		blockNumber, err := cv.getConfDeepBlockNumber(ctx)
+		if err != nil {
+			return [32]byte{}, fmt.Errorf("failed to get context block: %w", err)
+		}
+		return cv.retrieveBatchMetadataHash(ctx, batchID, blockNumber)
+	})
 	if err != nil {
-		return fmt.Errorf("failed to get batch metadata hash: %w", err)
-	}
-	if bytes.Equal(expectedHash[:], make([]byte, 32)) {
-		return ErrBatchMetadataHashNotFound
+		return fmt.Errorf("retrieving batch that was confirmed at block %v: %w", confirmationBlockNumber, err)
 	}
 
-	// 2. ensure that hash generated from local cert matches one stored on-chain
-	actualHash, err := HashBatchMetadata(header, recordHash, confirmationNumber)
+	// 3. compute the hash of the batch metadata received as argument
+	header := &binding.IEigenDAServiceManagerBatchHeader{
+		BlobHeadersRoot:       [32]byte(batchMetadata.GetBatchHeader().GetBatchRoot()),
+		QuorumNumbers:         batchMetadata.GetBatchHeader().GetQuorumNumbers(),
+		ReferenceBlockNumber:  batchMetadata.GetBatchHeader().GetReferenceBlockNumber(),
+		SignedStakeForQuorums: batchMetadata.GetBatchHeader().GetQuorumSignedPercentages(),
+	}
+	recordHash := [32]byte(batchMetadata.GetSignatoryRecordHash())
+	computedHash, err := HashBatchMetadata(header, recordHash, confirmationBlockNumber)
 	if err != nil {
 		return fmt.Errorf("failed to hash batch metadata: %w", err)
 	}
 
-	equal := slices.Equal(expectedHash[:], actualHash[:])
+	// 4. ensure that hash generated from local cert matches one stored on-chain
+	equal := slices.Equal(onchainHash[:], computedHash[:])
```

Comment:
knit - there is an equality func in utils: https://github.com/Layr-Labs/eigenda-proxy/blob/main/utils/utils.go#L31-L43

Reply:
Didn't know! But checked and that implementation is literally the same as [slices.Equal](https://pkg.go.dev/golang.org/x/exp/slices#Equal) though. I'd prefer sticking with stdlib (even if in exp package) stuff when we can.

Reply:
understood - lets cut an issue to remove that utils function entirely then